### PR TITLE
TST: GH23452 test reorder_categories() on categorical index

### LINKED ISSUE #23452: reorder_categories() on categorical index of interval values produces bizarre results
#### Problem description

Categorical indexes where the individual values are of type interval produce bizarre results when reordered with `.reorder_categories`.  This important because these indexes and columns are produced by `.cut`, so they occur in analysis even if the user never directly creates them.

#### Minimal Example 

With strings, this works as expected:
```python
d = pd.DataFrame(index=pd.CategoricalIndex(['a','b','c'],ordered=True), data={'x':[1,2,3]})
d.index=d.index.reorder_categories(d.index.categories[::-1])
print(d.sort_index())
```

```text
   x
c  3
b  2
a  1
```

With intervals, it does not:
```python
e=pd.DataFrame(index=pd.CategoricalIndex([pd.Interval(0,1), pd.Interval(1,2), pd.Interval(2,3)],ordered=True),data={'x':[1,2,3]})
e.index=e.index.reorder_categories(e.index.categories[::-1])
print(e.sort_index())
```

```text
     x
NaN  1
NaN  2
NaN  3
```

#### Expected Output

```text
        x
(2, 3]  3
(1, 2]  2
(0, 1]  1
```

#### More Realistic Example
```python
tips=sns.load_dataset('tips')
tips['total_bin']=pd.cut(tips['total_bill'],10)
tips.total_bin.cat.reorder_categories(tips.total_bin.cat.categories[::-1],inplace=True)
tips.head()
```

```text
   total_bill   tip     sex smoker  day    time  size total_bin
0       16.99  1.01  Female     No  Sun  Dinner     2       NaN
1       10.34  1.66    Male     No  Sun  Dinner     3       NaN
2       21.01  3.50    Male     No  Sun  Dinner     3       NaN
3       23.68  3.31    Male     No  Sun  Dinner     2       NaN
4       24.59  3.61  Female     No  Sun  Dinner     4       NaN
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.5.final.0
python-bits: 64
OS: Linux
OS-release: 3.16.0-4-amd64
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.23.0
pytest: 3.5.1
pip: 10.0.1
setuptools: 39.1.0
Cython: 0.28.2
numpy: 1.14.3
scipy: 1.1.0
pyarrow: None
xarray: None
IPython: 6.4.0
sphinx: 1.7.4
patsy: 0.5.0
dateutil: 2.7.3
pytz: 2018.4
blosc: None
bottleneck: 1.2.1
tables: 3.4.3
numexpr: 2.6.5
feather: None
matplotlib: 2.2.2
openpyxl: 2.5.3
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 1.0.4
lxml: 4.2.1
bs4: 4.6.0
html5lib: 1.0.1
sqlalchemy: 1.2.7
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by fokoid**
- [x] closes #23452
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Couldn't see entries for tests in whatsnew files so I assume it's not needed if only a test case is added This is my first Pandas contribution, looking forward to your feedback.

### CODE REVIEW

File: pandas/tests/indexing/test_categorical.py [hunk 1]
```diff
@@ -807,3 +807,31 @@ def test_loc_with_non_string_categories(self, idx_values, ordered):
...
+        "categories",
+        [
+            pytest.param(["a", "b", "c"], id="str"),
+            pytest.param(
+                [pd.Interval(0, 1), pd.Interval(1, 2), pd.Interval(2, 3)],
+                id="pd.Interval",
+            ),
+        ],
+    )
+    def test_categorical_index_reorder_categories(self, categories):
```
Comment (arw2019) [inline pandas/tests/indexing/test_categorical.py:821]: `test_reorder_index_with_categories`

File: pandas/tests/indexing/test_categorical.py [hunk 2]
```diff
@@ -807,3 +807,31 @@ def test_loc_with_non_string_categories(self, idx_values, ordered):
...
+        # GH23452
+        df = DataFrame(
+            {"foo": range(len(categories))},
+            index=CategoricalIndex(
+                data=categories, categories=categories, ordered=True
+            ),
+        )
+        df.index = df.index.reorder_categories(df.index.categories[::-1])
+        actual = df.sort_index()
```
Comment (arw2019) [inline pandas/tests/indexing/test_categorical.py:830]: call it `result`

**Review by arw2019 — CHANGES_REQUESTED** [review]
Thanks @fokoid for the PR!

Looks good, some comments

File: pandas/tests/indexing/test_categorical.py [hunk 3]
```diff
@@ -807,3 +807,31 @@ def test_loc_with_non_string_categories(self, idx_values, ordered):
...
         tm.assert_frame_equal(result, expected)
+
+    @pytest.mark.parametrize(
+        "categories",
+        [
+            pytest.param(["a", "b", "c"], id="str"),
+            pytest.param(
+                [pd.Interval(0, 1), pd.Interval(1, 2), pd.Interval(2, 3)],
+                id="pd.Interval",
+            ),
```
Comment (dsaxton) [inline pandas/tests/indexing/test_categorical.py:818]: Seems these could simply be lists without the pytest.param?

Comment (fokoid) [reply pandas/tests/indexing/test_categorical.py:818]: Using `pytest.param` with the `id` parameter results in clearer test names:
```
test_categorical.py::TestCategoricalIndex::test_categorical_index_reorder_categories[str] 
test_categorical.py::TestCategoricalIndex::test_categorical_index_reorder_categories[pd.Interval] 
```
Using plain lists results in:
```
test_categorical.py::TestCategoricalIndex::test_categorical_index_reorder_categories[categories0] 
test_categorical.py::TestCategoricalIndex::test_categorical_index_reorder_categories[categories1] 
```

File: pandas/tests/indexing/test_categorical.py (hunk 1 above)
Comment (fokoid) [reply pandas/tests/indexing/test_categorical.py:821]: Thanks. Is there a rule (or at least some rough guidance) that would help me choose better test names next time? I couldn't see anything in the style guide, but would be happy to update it with any accepted conventions.

Comment (arw2019) [reply pandas/tests/indexing/test_categorical.py]: No rules but we want it to be as easy as possible to figure out what the test does if/when if fails on a future PR

**Comment by fokoid** [comment]
Requested changes have been applied, is this good to merge now?

**Review by dsaxton — APPROVED** [review]

**Comment by dsaxton** [comment]
Thanks @fokoid

**Comment by fokoid** [comment]
Sorry to bother you guys - could you please mark this PR with `hacktoberfest-accepted` as per #36837?

**Pull request merged**